### PR TITLE
Conversion to TypeScript

### PR DESCRIPTION
**Pull request by dgreene1**
This fixes https://github.com/lissitz/react-accessible-treeview/issues/23

@lissitz, we were extremely diligent about not changing any logic so that this PR was nothing but JS -> TS. To make this PR easier to follow, please approve & merge https://github.com/lissitz/react-accessible-treeview/pull/27 first.

After this is done, we (@mellis481 and I) would like to volunteer as maintainers so we can make sure that this release gets published and that we can continue to make quality of life improvements.

I will comment in this PR about some things that might seem like suspicious changes but that follow standard practices so that it's easier for you to review.

Nice work @mellis481 and our team for this impressive PR! :)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,6 +1,7 @@
 .DS_Store
 node_modules
 dist
-package-lock.json
+website/package-lock.json
```
Comment (dgreene1) [inline .gitignore:4]: We made this change because:

1. you can't gitignore the package-lock.json because it will prevent `npm ci` from working which means that published versions will not have locked transitive dependencies
2. you do want to gitignore the lockfile for the storybook folder since that's not deployed.

File: .gitignore [hunk 2]
```diff
@@ -1,6 +1,7 @@
...
 styleguide
-yarn.lock
\ No newline at end of file
+website/yarn.lock
+storybook-static
```
Comment (dgreene1) [inline .gitignore:7]: This is a build artifact, and build artifacts are not to be checked in.

File: .travis.yml [hunk 3]
```diff
@@ -1,8 +1,8 @@
 language: node_js
 node_js:
-  - "11"
-  - "10"
-  - "8"
+  - "16.15.0"
```
Comment (dgreene1) [inline .travis.yml:3]: This fixes the Travis errors that had prevented all PRs from merging. We're now only checking this on LTS

File: jest.config.js [hunk 4]
```diff
@@ -1,10 +1,16 @@
 module.exports = {
   clearMocks: true,
   coverageDirectory: "coverage",
-  moduleFileExtensions: ["js", "jsx"],
+  preset: "ts-jest",
   moduleDirectories: ["node_modules"],
   moduleNameMapper: {
     "\\.(css|less)$": "<rootDir>/__mocks__/styleMock.js",
-    "\\.(gif|ttf|eot|svg)$": "<rootDir>/__mocks__/fileMock.js"
```
Comment (dgreene1) [inline jest.config.js:8]: This wasn't being used.

**Comment by lissitz** [comment]
Hi. Thank you for your contribution. I've added you both as collaborators, so you can merge the PR.

File: src/TreeView/utils.ts [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+  );
+
+const isIE = () => window.navigator.userAgent.match(/Trident/);
+
+export const onComponentBlur = (
+  event: React.FocusEvent,
+  treeNode: HTMLUListElement | null,
+  callback: () => void
+) => {
+  if (treeNode == null) {
```
Comment (mellis481) [inline src/TreeView/utils.ts:279]: @lissitz It's possible to have a null ref passed to this function for the `treeNode` prop. I'm logging a warning to the console unless there is something more preferable.

CC: @dgreene1

File: src/TreeView/index.tsx [hunk 6]
```diff
@@ -826,41 +1105,41 @@ const handleKeyDown = ({
       case "ArrowUp": {
         event.preventDefault();
         const previous = getPreviousAccessible(data, id, expandedIds);
-        if (previous != null || !disabledIds.has(previous)) {
+        if (previous != null && !disabledIds.has(previous)) {
```
Comment (mellis481) [inline src/TreeView/index.tsx:1108]: @dgreene1 Can you validate this change and the one on 1129 is good please?

Comment (dgreene1) [reply src/TreeView/index.tsx:1108]:
```suggestion
        if (previous != null || (previous == null && !disabledIds.has(previous))) {
```
I'd have to do testing, but this gross suggestion I'm making is the most literal translation of the JS version. Like this should be have the same way that it did before.

The issue with your first pass (`previous != null && !disabledIds.has(previous)`) is that it doesn't allow nulls in where `previous` is `null` and `disabledIds` has null. But since you changed the type to enforce that `disabledIds` is now `Set<number>` (when it was previously `Set<number | null>` then I think both your work here or my suggestion have the same ultimate result.

tl;dr: you're good to go either way.

File: src/TreeView/index.tsx [hunk 7]
```diff
@@ -742,18 +1009,16 @@ const NodeGroup = ({
...
     {expandedIds.has(element.id) &&
-      element.children.map((x, index) => (
+      element.children.map((x) => (
         <Node
           data={data}
           expandedIds={expandedIds}
           baseClassNames={baseClassNames}
           key={x}
           element={data[x]}
-          setsize={element.children.length}
```
[the remark this answers is not in the document]
Comment (dgreene1) [reply src/TreeView/index.tsx:755]: FYI @lissitz this was fixed in https://github.com/lissitz/react-accessible-treeview/pull/31/commits/f7c275c5618a90a17d067d567a809c52472b80f6 and I even included test automation. yay! 🌟
